cli_tools/common/utils/daisyutils/daisy_utils.go

```diff
@@ -429,3 +397,96 @@ func GetDiskURI(project, zone, name string) string {
 func GetInstanceURI(project, zone, name string) string {
 	return fmt.Sprintf("projects/%v/zones/%v/instances/%v", project, zone, name)
 }
+
+// ParseWorkflow parses Daisy workflow file and returns Daisy workflow object or error in case of failure
+func ParseWorkflow(path string, varMap map[string]string, project, zone, gcsPath, oauth, dTimeout, cEndpoint string, disableGCSLogs, disableCloudLogs, disableStdoutLogs bool) (*daisy.Workflow, error) {
+	w, err := daisy.NewFromFile(path)
+	if err != nil {
+		return nil, err
+	}
+Loop:
+	for k, v := range varMap {
+		for wv := range w.Vars {
+			if k == wv {
+				w.AddVar(k, v)
+				continue Loop
+			}
+		}
+		return nil, daisy.Errf("unknown workflow Var %q passed to Workflow %q", k, w.Name)
+	}
+
+	EnvironmentSettings{
+		Project:           project,
+		Zone:              zone,
+		GCSPath:           gcsPath,
+		OAuth:             oauth,
+		Timeout:           dTimeout,
+		ComputeEndpoint:   cEndpoint,
+		DisableGCSLogs:    disableGCSLogs,
+		DisableCloudLogs:  disableCloudLogs,
+		DisableStdoutLogs: disableStdoutLogs,
+	}.ApplyToWorkflow(w)
+
+	return w, nil
+}
+
+// EnvironmentSettings controls the resources that are used during tool execution.
+type EnvironmentSettings struct {
+	// Location of workflows
+	WorkflowDirectory string
+
+	// Fields from daisy.Workflow
+	Project, Zone, GCSPath, OAuth, Timeout, ComputeEndpoint string
+	DisableGCSLogs, DisableCloudLogs, DisableStdoutLogs     bool
+
+	// An optional prefix to include in the bracketed portion of daisy's stdout logs.
+	// Gcloud does a prefix match to determine whether to show a log line to a user.
+	//
+	// With a prefix of `disk-1`, for example, the workflow in `importer.NewDaisyInflater`
+	// emits log messages starting with `[disk-1-inflate]`.
+	DaisyLogLinePrefix string
+
+	// Worker instance customizations
+	Network, Subnet       string
+	ComputeServiceAccount string
+	NoExternalIP          bool
+	Labels                map[string]string
+	ExecutionID           string
+	StorageLocation       string
+}
+
+// ApplyToWorkflow sets fields on daisy.Workflow from the environment settings.
+func (env EnvironmentSettings) ApplyToWorkflow(w *daisy.Workflow) {
+	w.Project = env.Project
+	w.Zone = env.Zone
+	if env.GCSPath != "" {
+		w.GCSPath = env.GCSPath
+	}
+	if env.OAuth != "" {
+		w.OAuthPath = env.OAuth
+	}
+	if env.Timeout != "" {
+		w.DefaultTimeout = env.Timeout
+	}
+	if env.ComputeEndpoint != "" {
+		w.ComputeEndpoint = env.ComputeEndpoint
+	}
+	if env.DisableGCSLogs {
+		w.DisableGCSLogging()
+	}
+	if env.DisableCloudLogs {
+		w.DisableCloudLogging()
+	}
+	if env.DisableStdoutLogs {
+		w.DisableStdoutLogging()
+	}
+}
+
+// UpdateAllInstanceNoExternalIP updates all Create Instance steps in the workflow to operate
+// when no external IP access is allowed by the VPC Daisy workflow is running in.
+func UpdateAllInstanceNoExternalIP(workflow *daisy.Workflow, noExternalIP bool) {
```

Comment:
Why is no external IP update done from here, but also in NewDaisyWorker?

Btw, it's a bit hard to follow such a large PR. When a lot of code is renamed, moved, it's a better idea to do that step in one PR then add/change functionality in another.

Reply:
Good question -- that was for refactoring -- the code from `UpdateAllInstanceNoExternalIP` was copied into `RemoveExternalIPTraversal`, and then `UpdateAllInstanceNoExternalIP` called `RemoveExternalIPTraversal`.

Also, in retrospect, I agree this would have been clearer with the split you suggested. Sorry about that; I'll do better in the future!